### PR TITLE
fix(auth): Moving credential provider to main

### PR DESCRIPTION
**Pull request by gpanshu**
- [x] PR title and description conform to [Pull Request](https://github.com/aws-amplify/amplify-android/blob/main/CONTRIBUTING.md#pull-request-guidelines) guidelines.

*Issue #, if available:*

*Description of changes:*

*How did you test these changes?*
(Please add a line here how the changes were tested)

*Documentation update required?*
- [x] No
- [ ] Yes (Please include a PR link for the documentation update)

*General Checklist*
- [ ] Added Unit Tests
- [ ] Added Integration Tests
- [x] Security oriented best practices and standards are followed (e.g. using input sanitization, principle of least privilege, etc)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: aws-auth-cognito/src/test/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPluginTest.kt [hunk 1]
```diff
@@ -1033,7 +1033,7 @@ class RealAWSCognitoAuthPluginTest {
             onError
         )
 
-        assertTrue { listenLatch.await(5, TimeUnit.SECONDS) }
+        assertTrue { listenLatch.await(5, TimeUnit.MINUTES) }
```
Comment (div5yesh) [inline aws-auth-cognito/src/test/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPluginTest.kt:1036]: Seconds.

File: core/src/main/java/com/amplifyframework/auth/CognitoCredentialsProvider.kt [hunk 2]
```diff
@@ -68,11 +65,34 @@ internal open class CognitoCredentialsProvider : CredentialsProvider {
...
+    }
+
     fun getAccessToken(onResult: Consumer<String>, onFailure: Consumer<Exception>) {
         Amplify.Auth.fetchAuthSession(
             { session ->
                 val tokens = (session as? AWSCognitoAuthSession)?.userPoolTokensResult?.value?.accessToken
-                tokens?.let { onResult.accept(it) }
+                tokens?.let { onResult.accept(tokens) }
                     ?: onFailure.accept(
                         ApiException.ApiAuthException(
```
Comment (div5yesh) [inline core/src/main/java/com/amplifyframework/auth/CognitoCredentialsProvider.kt:97]: Amplify or Auth Exception.

File: core/src/main/java/com/amplifyframework/auth/CognitoCredentialsProvider.kt [hunk 3]
```diff
@@ -68,11 +65,34 @@ internal open class CognitoCredentialsProvider : CredentialsProvider {
...
+     * Request identityId from the provider.
+     */
+    override suspend fun getIdentityId(): String {
+        return suspendCoroutine { continuation ->
+            Amplify.Auth.fetchAuthSession(
+                { authSession ->
+                    authSession.toAWSCognitoAuthSession()?.identityIdResult?.value?.let {
+                        continuation.resume(it)
+                    } ?: continuation.resumeWithException(
+                        Exception(
```
Comment (div5yesh) [inline core/src/main/java/com/amplifyframework/auth/CognitoCredentialsProvider.kt:78]: Amplify or Auth Exception.

File: core/src/main/java/com/amplifyframework/auth/AWSAuthSessionInternal.kt [hunk 4]
```diff
@@ -0,0 +1,26 @@
+package com.amplifyframework.auth
+
+/*
```
Comment (div5yesh) [inline core/src/main/java/com/amplifyframework/auth/AWSAuthSessionInternal.kt:3]: nit: header goes before package name

File: core/src/main/java/com/amplifyframework/auth/CognitoCredentialsProvider.kt [hunk 5]
```diff
@@ -76,10 +76,28 @@ internal class CognitoCredentialsProvider : AuthCredentialsProvider {
...
+        )
+    }
 }
 
-private fun AuthSession.toAWSCognitoAuthSession(): AWSCognitoAuthSession? {
-    if (this is AWSCognitoAuthSession) {
+private fun AuthSession.toAWSAuthSession(): AWSAuthSessionInternal? {
+    if (this is AWSAuthSessionInternal) {
         return this
     }
```
Comment (div5yesh) [inline core/src/main/java/com/amplifyframework/auth/CognitoCredentialsProvider.kt:103]:
```suggestion
    return as? AWSAuthSessionInternal

```

File: core/src/main/java/com/amplifyframework/auth/CognitoCredentialsProvider.kt [hunk 6]
```diff
@@ -13,37 +13,36 @@
...
 import kotlin.coroutines.resumeWithException
 import kotlin.coroutines.suspendCoroutine
 
 /**
- * Internal implementation of cognito credentials provider.
- * This will be ported to core once it seems feasible to do so.
+ * Wrapper to provide credentials from Auth synchronously and asynchronously
  */
-internal class CognitoCredentialsProvider : AuthCredentialsProvider {
+open class CognitoCredentialsProvider : CredentialsProvider, AuthCredentialsProvider {
```
Comment (div5yesh) [inline core/src/main/java/com/amplifyframework/auth/CognitoCredentialsProvider.kt:30]:
```suggestion
open class CognitoCredentialsProvider : AuthCredentialsProvider {
```

File: core/src/main/java/com/amplifyframework/auth/CognitoCredentialsProvider.kt [hunk 7]
```diff
@@ -76,10 +76,28 @@ internal class CognitoCredentialsProvider : AuthCredentialsProvider {
             )
         }
     }
+
+    fun getAccessToken(onResult: Consumer<String>, onFailure: Consumer<Exception>) {
```
Comment (div5yesh) [inline core/src/main/java/com/amplifyframework/auth/CognitoCredentialsProvider.kt:80]: Which category needs `getAccessToken`? Consider adding this function to `AuthCredentialsProvider` interface too or remove it if not used.

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/AWSCognitoAuthSession.kt [hunk 8]
```diff
@@ -92,7 +59,38 @@ internal fun AmplifyCredential.isValid(): Boolean {
 
 internal fun AmplifyCredential.getCognitoSession(
     exception: AuthException = SignedOutException()
-): AWSCognitoAuthSession {
+): AWSAuthSessionInternal {
```
Comment (div5yesh) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/AWSCognitoAuthSession.kt:62]:
```suggestion
): AWSCognitoAuthSession {
```

**Review by div5yesh — DISMISSED** [review]
Few comments, otherwise looks good.

File: core/src/main/java/com/amplifyframework/auth/CognitoCredentialsProvider.kt (hunk 7 above)
Comment (gpanshu) [reply core/src/main/java/com/amplifyframework/auth/CognitoCredentialsProvider.kt:80]: API category needs it. I will add it to the interface.

**Review by div5yesh — APPROVED** [review]

**Review by banji180 — APPROVED** [review]

**Pull request merged**